playbooks/roles/mattermail/tasks/main.yml

```diff
@@ -0,0 +1,49 @@
+- name: create mattermail group
+  group:
+    name: mattermail
+
+- name: create mattermail user
+  user:
+    name: mattermail
+    group: mattermail
+    home: "{{ MATTERMAIL_HOME }}"
+    system: yes
+
+- name: Install git and golang packages
+  apt:
+    name:
+      - git
+      - golang
+    update_cache: yes
+
+- name: build Mattermail binary
+  command: go get {{ MATTERMAIL_REPO }}
+  args:
+    creates: "{{ MATTERMAIL_HOME }}/bin/mattermail"
```

Comment:
Ah, that's right. But I still don't think we should be re-building the binary each time, that may cause issues for a running Mattermail, or maybe not? [This section from the Mattermost role](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/mattermost/tasks/mattermost.yml#L40-L106) may provide some help in what to do here. It's kind of hacky, and the other way is to use Ansible facts, but I'm not sure how much time we have to implement that.

Reply:
One way is to store the git commit id in a file in MATTERMAIL_HOME and compare that to the current commit it to decide if an update it needed. It doesn't need to be a complex thing, you can use use ``set_fact`` I think.

Reply:
I didn't expect this to be used for upgrading as well. There shouldn't be any issue replacing the binary while Mattermail is running, but the service should be restarted instead of enabled and started.
I'll evaluate storing the git hash anyway.

Reply:
Enabling is idempotent if it's already enabled nothing will happen. I don't think we will be deploying a lot of new Matermost servers, we'll probably be running the playbook on the same server again and again. Ideally, the upgrades would be controlled using some variable with the version number, like @UmanShahzad  suggested above with the binary url.

Reply:
You can add a `notify: restart mattermail` and a handler to restart it on this step or whatever seems right, so the systemd service will get restarted.

- [Example notify](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/consul/tasks/main.yml#L48)
- [Example handler](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/consul/handlers/main.yml)

Reply:
@xitij2000 The idea of having our own repository for Mattermail would be that we control the master branch (which would be used to build the binary using `go get`)  thus removing any responsibility from the playbook to handle versions, simplifying configuration. What do you think?

@UmanShahzad It seems that having it always enable and restart at the end also works and seems simpler:
```yaml
- name: start and enable the mattermail service
  service:
    name: mattermail
    state: restarted
    enabled: yes
```

Reply:
Sure, having our own fork is OK to if we don't go for the binary approach.

Reply:
About having a notifier, that makes sense but only if we're only running the update if there is a version upgrade. Otherwise it will always run the handler so might as well just have a regular task.

Reply:
@viadanna The notifier would only get run if the task got run successfully, and in general it's not good practice to restart the service if it isn't needed; imagine doing this for MySQL, for example, it would restart the service and bring down all connections anytime we run the playbook, which isn't ideal. I agree it isn't that critical here, but we should still be consistent / follow best practices.

Reply:
@xitij2000 Yep, I would only say add the notifier if we plan on implementing the ability to skip the install step if it isn't needed. Otherwise it would indeed always get run and would make no difference.